ext/node/zlib/mod.rs

```diff
@@ -0,0 +1,474 @@
+    match self.mode {
+      DEFLATE | GZIP | DEFLATERAW => {
+        self.err = unsafe { libz_sys::deflate(&mut self.strm, flush) };
+      }
+      UNZIP if self.strm.avail_in > 0 => {
+        let mut next_expected_header_byte = Some(0);
+
+        if self.gzib_id_bytes_read == 0 {
+          let byte = unsafe { *self.strm.next_in.offset(0) };
+          if byte == GZIP_HEADER_ID1 {
+            self.gzib_id_bytes_read = 1;
+            next_expected_header_byte = Some(1);
+
+            // Not enough.
+            if self.strm.avail_in == 1 {
+              self.write_in_progress = false;
+              return Ok(());
+            }
+          } else {
+            self.mode = INFLATE;
+            next_expected_header_byte = None;
+          }
+        }
+
+        if self.gzib_id_bytes_read == 1 && next_expected_header_byte.is_some() {
+          let byte = unsafe {
+            *self.strm.next_in.offset(next_expected_header_byte.unwrap())
+          };
+          if byte == GZIP_HEADER_ID2 {
+            self.gzib_id_bytes_read = 2;
+            self.mode = GUNZIP;
+          } else {
+            self.mode = INFLATE;
+          }
+        } else if next_expected_header_byte != None {
+          return Err(type_error(
+            "invalid number of gzip magic number bytes read",
+          ));
+        }
+      }
+      _ => {}
+    }
+
+    match self.mode {
+      INFLATE | GUNZIP | INFLATERAW => {
+        self.err = unsafe { libz_sys::inflate(&mut self.strm, self.flush) };
+
+        // TODO(@littledivy): Use if let chain when it is stable.
+        // https://github.com/rust-lang/rust/issues/53667
+        if self.err == Z_NEED_DICT && self.dictionary.is_some() {
+          // Data was encoded with dictionary
+          let dictionary = self.dictionary.as_ref().unwrap();
+          self.err = unsafe {
+            libz_sys::inflateSetDictionary(
+              &mut self.strm,
+              dictionary.as_ptr() as _,
+              dictionary.len() as _,
+            )
+          };
+
+          if self.err == Z_OK {
+            self.err = unsafe { libz_sys::inflate(&mut self.strm, self.flush) };
+          } else if self.err == Z_DATA_ERROR {
+            self.err = Z_NEED_DICT;
+          }
+        }
+
+        while self.strm.avail_in > 0
+          && self.mode == GUNZIP
+          && self.err == Z_STREAM_END
+          && unsafe { *self.strm.next_in as u8 } != 0x00
+        {
+          self.err = unsafe { libz_sys::inflateReset(&mut self.strm) };
+          self.err = unsafe { libz_sys::inflate(&mut self.strm, self.flush) };
+        }
+      }
+      _ => {}
+    }
+    if self.err == Z_BUF_ERROR
+      && !(self.strm.avail_out != 0 && self.flush == Z_FINISH)
+    {
+      self.err = Z_OK;
+    };
+    self.write_in_progress = false;
+    Ok(())
+  }
+}
+
+struct Zlib {
+  inner: RefCell<ZlibInner>,
+}
+
+impl deno_core::Resource for Zlib {
+  fn name(&self) -> Cow<str> {
+    "zlib".into()
+  }
+}
+
+const NONE: i32 = 0;
```

Comment:
Can this be an enum with a `#[repr(u8)]`?

Reply:
Good point, will do - this should simplify the `match` since its exhaustive